src/states/elder/mod.rs

```diff
@@ -665,178 +656,151 @@ impl Elder {
+    fn handle_backlogged_message(
         &mut self,
-        signed_msg: SignedRoutingMessage,
+        sender: Option<ConnectionInfo>,
+        msg: Message,
     ) -> Result<(), RoutingError> {
-        trace!(
-            "{} - Handle backlogged signed message: {:?}",
-            self,
-            signed_msg.routing_message()
-        );
+        trace!("{} - Handle backlogged message: {:?}", self, msg);
 
-        if self.in_dst_location(&signed_msg.routing_message().dst) {
-            if !self
-                .verify_signed_message(&signed_msg)
-                .map(|trusted| trusted)
-                .unwrap_or(false)
-            {
+        if self.in_dst_location(&msg.dst) {
+            if !self.is_message_verified(&msg) {
                 return Ok(());
             }
 
             // If message still for us and we still trust it, then it must not be stale.
-            self.update_our_knowledge(&signed_msg);
-            self.routing_msg_queue.push_back(signed_msg);
+            self.update_our_knowledge(&msg);
+            self.msg_queue.push_back(msg.into_queued(sender));
         }
 
         Ok(())
     }
 
-    fn verify_signed_message(&self, msg: &SignedRoutingMessage) -> Result<bool, RoutingError> {
-        match msg.verify(self.chain.get_their_key_infos()) {
-            Ok(VerifyStatus::Full) => Ok(true),
-            Ok(VerifyStatus::ProofTooNew) if msg.routing_message().dst.is_multiple() => {
-                // Proof is too new which can only happen if we've been already demoted but are
-                // lagging behind (or the sender is faulty/malicious). We can't handle the
-                // message ourselves but the other elders likely can.
-                Ok(false)
-            }
-            Ok(VerifyStatus::ProofTooNew) => Err(RoutingError::UntrustedMessage),
-            Err(error) => Err(error),
-        }
-    }
-
-    fn dispatch_routing_message(
+    fn dispatch_message(
         &mut self,
-        signed_msg: SignedRoutingMessage,
+        sender: Option<ConnectionInfo>,
+        msg: Message,
         outbox: &mut dyn EventBox,
     ) -> Result<Transition, RoutingError> {
-        use crate::messages::Variant::*;
-
-        let (msg, src_authority) = signed_msg.into_parts();
-
-        match msg.content {
-            UserMessage { .. } => (),
-            _ => trace!("{} Got routing message {:?}.", self, msg),
+        match msg.variant {
+            Variant::UserMessage { .. } => (),
+            _ => trace!("{} Got {:?}.", self, msg),
         }
 
-        match (msg.content, msg.src, msg.dst) {
-            (
-                NeighbourInfo(elders_info),
-                src @ SrcLocation::Section(_),
-                dst @ DstLocation::PrefixSection(_),
-            ) => {
-                self.handle_neighbour_info(elders_info, src, dst, src_authority)?;
-                Ok(Transition::Stay)
+        match msg.variant {
+            Variant::NeighbourInfo(elders_info) => {
```

Comment:
I think this is not right, we still need to do that same match we did before.
i.e still want to check the sending authority and the destination are as expected.

Reply:
So I see you've added `let _ = msg.src.as_section()?;..` But it is only for this, that comment applied probably to all messages. I'm still thinking matching on the tuple make sense?

Reply:
But I guess it makes exhaustive matching not possible. Not sure what is better.

Reply:
> But it is only for this, that comment applied probably to all messages

But I do that for all messages. If you notice there are calls like `as_section`, `as_node`, etc... In every branch except `UserMessage` because that one can be sent with any src/dst. Not sure matching on the tuple would buy as much. And as you said, we would lose exhaustive match.

Reply:
Indeed, missed that.